src/Storages/MergeTree/MergeTreeCurrentlyRestoringFromBackup.cpp

```diff
@@ -0,0 +1,333 @@
+    void checkNoPartsExist(Int64 base_increment_value, const std::optional<AllocatedBlockNumbers> & ignore_block_numbers, CheckForNoPartsReason reason) const
+    {
+        if (reason == CheckForNoPartsReason::NONE)
+            return;
+
+        LOG_INFO(log, "Checking that no parts exist in the table before restoring its data (reason: {})", magic_enum::enum_name(reason));
+
+        /// The order of checks is important:
+        /// an INSERT command first allocates a block number, then it attaches it to the table.
+        /// That's why it's more reliable to check first block numbers, then data parts.
+
+        auto allocated_block_number = findAnyAllocatedBlockNumber(base_increment_value, ignore_block_numbers);
+
+        auto part_name_in_storage = findAnyPartInStorage();
+        if (part_name_in_storage)
+            throwTableIsNotEmpty(reason, fmt::format("part {} exists", *part_name_in_storage));
+
+        auto restoring_part_name = findAnyRestoringPart();
+        if (restoring_part_name)
+            throwTableIsNotEmpty(reason, fmt::format("concurrent RESTORE is creating part {}", *restoring_part_name));
+
+        if (allocated_block_number)
+            throwTableIsNotEmpty(reason, fmt::format("concurrent INSERT or ATTACH is using block number {}", *allocated_block_number));
+    }
+
+    /// Finds any part in the storage.
+    std::optional<String> findAnyPartInStorage() const
+    {
+        if (storage.getTotalActiveSizeInBytes() == 0)
+            return {};
+
+        auto parts_lock = storage.lockParts();
+
+        auto parts = storage.getDataPartsVectorForInternalUsage({MergeTreeDataPartState::Active, MergeTreeDataPartState::PreActive}, parts_lock);
+        if (parts.empty())
+            return {};
+
+        return parts.front()->name;
+    }
+
+    Int64 getCurrentIncrementValue() const
+    {
+        return storage.increment.value.load();
+    }
+
+    /// Finds any locked block number excluding ones we have just allocated.
+    std::optional<Int64> findAnyAllocatedBlockNumber(Int64 base_increment_value, const std::optional<AllocatedBlockNumbers> & ignore_block_numbers) const
+    {
+        Int64 base = base_increment_value;
+        Int64 current = getCurrentIncrementValue();
+
+        if (current < base)
+        {
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "The block number increment changed in a unexpected way: base={}, current={}",
+                            base, current);
+        }
+
+        if (!ignore_block_numbers)
+        {
+            if (current > base)
+                return base + 1;
+            return {};
+        }
+
+        Int64 ignore_first = ignore_block_numbers->first;
+        Int64 ignore_count = ignore_block_numbers->count;
+
+        if ((ignore_first < base + 1) || (ignore_first + ignore_count < ignore_first) || (current < ignore_first + ignore_count - 1))
```

Comment:
Isn't `ignore_first + ignore_count < ignore_first` equivalent to `ignore_count < 0`? Or did you mean smth else? Btw, why is the count signed?

Reply:
I've fixed this check.